lib/TPP/LowerPacksAndUnpacks.cpp

```diff
@@ -0,0 +1,254 @@
+    int64_t upTo = packingOp.getDestType().getRank() - 2;
+    return SmallVector<int64_t>(upTo, 1);
+  }
+  return llvm::to_vector(packingOp.getStaticTiles());
+}
+
+// Tile`operation` using `tileSizes`, the loops are annotated with
+// `kLoopParallel` and later converted to scf.forall.
+static FailureOr<scf::SCFTilingResult> tileOp(RewriterBase &rewriter,
+                                              TilingInterface operation,
+                                              ArrayRef<int64_t> tileSizes) {
+  auto options = scf::SCFTilingOptions().setTileSizes(
+      getAsIndexOpFoldResult(rewriter.getContext(), tileSizes));
+  FailureOr<scf::SCFTilingResult> tilingResult =
+      scf::tileUsingSCFForOp(rewriter, operation, options);
+  if (failed(tilingResult))
+    return failure();
+  if (!tilingResult->loops.empty()) {
+    tilingResult->loops[0]->setAttr(
+        linalgx::utils::kLoopParallel,
+        rewriter.getStringAttr(linalgx::utils::kLoopRoot));
+  }
+  return tilingResult;
+}
+
+// Fuse producer and consumer pack. Standalone packs are tiled into 2d tiles.
+static void fuseOrTilePacks(RewriterBase &rewriter, FunctionOpInterface func) {
+  SmallVector<tensor::PackOp> consumerPackOps;
+  SmallVector<tensor::PackOp> otherPacks;
+  SmallVector<tensor::UnPackOp> unPacks;
+  func->walk<WalkOrder::PostOrder>([&](tensor::PackOp consumerPackOp) {
+    Value source = consumerPackOp.getSource();
+    tensor::PackOp producerPackOp =
+        dyn_cast_or_null<tensor::PackOp>(source.getDefiningOp());
+    if (producerPackOp)
+      consumerPackOps.push_back(consumerPackOp);
+    else
+      otherPacks.push_back(consumerPackOp);
+  });
+  func->walk<WalkOrder::PostOrder>(
+      [&](tensor::UnPackOp unpackOp) { unPacks.push_back(unpackOp); });
+
+  // Tile and fuse.
+  for (auto consumerPackOp : consumerPackOps) {
+    // Step 2. Tile the operation.
+    SmallVector<int64_t> tileSizes =
+        getTileSizes(consumerPackOp, /*isConsumer=*/true);
+    if (tileSizes.empty())
+      continue;
+    auto tilingResult =
+        tileOp(rewriter, cast<TilingInterface>(consumerPackOp.getOperation()),
+               tileSizes);
+    if (failed(tilingResult))
+      continue;
+    auto tiledPack = dyn_cast<tensor::PackOp>(tilingResult->tiledOps.back());
+    assert(tiledPack);
+    // Step 3. Fuse consumer and producer.
+    auto forLoops =
+        llvm::to_vector(llvm::map_range(tilingResult->loops, [](Operation *op) {
+          return cast<scf::ForOp>(op);
+        }));
+    std::optional<scf::SCFFuseProducerOfSliceResult> fusedProducer =
+        scf::tileAndFuseProducerOfSlice(
+            rewriter,
+            cast<tensor::ExtractSliceOp>(tiledPack.getSource().getDefiningOp()),
+            forLoops);
+    if (!fusedProducer)
+      continue;
+    rewriter.replaceOp(consumerPackOp, tilingResult->replacements);
+  }
+
+  // Tile packs.
+  for (auto packOp : otherPacks) {
+    SmallVector<int64_t> tileSizes = getTileSizes(packOp);
```

Comment:
Shouldn't you be using the same tiles as the consumer? There doesn't seem to be any guarantee to this here or in `getTileSizes`.

Reply:
Here I don't have any consumer. All the packs are standalone.